lib/ecto/type.ex

```diff
@@ -644,45 +644,48 @@ defmodule Ecto.Type do
       :error
 
   """
-  @spec cast(t, term) :: {:ok, term} | {:error, keyword()} | :error
-  def cast({:parameterized, type, params}, value), do: type.cast(value, params)
-  def cast({:in, _type}, nil), do: :error
-  def cast(_type, nil), do: {:ok, nil}
-
-  def cast({:maybe, type}, value) do
+  @spec cast(t, term) :: {:ok, term} | {:ok, term, boolean} | {:error, keyword()} | :error
+  def cast(type, value, current \\ nil)
+  def cast({:parameterized, type, params}, value, current), do: type.cast_with_current(value, current, params)
+  def cast({:embed, embedded}, value, current), do: Ecto.Embedded.cast_with_current(value, current, embedded)
```

Comment:
This should be handled in `Ecto.Changeset`, as there is no such thing as `{:embed, embedded}` from the `Ecto.Type` perspective.

Reply:
For example, in `Ecto.Changeset.type!`, you can rewrite {:embed, embedded} into the parameterized one. :)